BUG: xarray data access

jklenzing opened the pull request:
# Description

Addresses #471

- Fixes a bug when attempting to set data by indices for a multi-dimensional xarray object. 
- Includes `pysat_testing2d_xarray` as part of standard data tests.
  - updated `test_setting_partial_data_by_*` tests with parametrize

Previous code assumes only epoch axis, and converts index values to timestamps.  However, if a secondary axis is used (Altitude in the case below), it will try to load altitude values as index values.  This bypasses the conversion and uses the integers directly as indices in this case.  This is the most flexible, as many types of dimensions can be used.

## Type of change

- Bug fix (non-breaking change which fixes an issue)

# How Has This Been Tested?

- Load Instrument
```
import numpy as np
import pysat
import pysatNASA
inst = pysat.Instrument(inst_module=pysatNASA.instruments.icon_mighti, tag='vector_wind_green', clean_level='none')
```
- Get data if you don't have it
```
import datetime as dt
inst.download(dt.datetime(2020, 1, 2), dt.datetime(2020, 1, 2)
```
- Try to set some values to nans
```
inst.load(2020, 2)
idx, idy, = np.where(inst['Wind_Quality'] != 1)
inst[idx, idy, 'Zonal_Wind'] = np.nan
```
In develop-3, this gives an error (see #471 for details)

**Test Configuration**:
*Mac OS X 10.15.7
* Python 3.8.3

# Checklist:

- [x] Make sure you are merging into the ``develop`` (not ``main``) branch
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [x] Add a note to ``CHANGELOG.md``, summarizing the changes

jklenzing commented:
Putting this up as draft for now as the current unit tests will not catch this bug.  While we test both pandas and xarray objects, only 1d data sets are actually tested in the current writeup.  This requires a rewrite (or duplication under another class) of all tests under test_instrument.py::TestBasics starting with either "test_setting_data_by" or "test_data_access_by" (14 tests total).

jklenzing commented:
Updated tests based on discussions with @aburrell.  The bug has been covered.

aburrell commented:
Ok, I got your test to pass (ish) and also broke it!  After downloading data I did:

```
print(inst.data)

<xarray.Dataset>
Dimensions:                               (Epoch: 2168, ICON_L22_Altitude: 84, N_Flags: 34)
Coordinates:
  * Epoch                                 (Epoch) float64 1.578e+12 ... 1.578e+12
  * ICON_L22_Altitude                     (ICON_L22_Altitude) float64 88.38 ... 304.6
Dimensions without coordinates: N_Flags
Data variables:
    Epoch_Full                            (Epoch, ICON_L22_Altitude) float64 ...
    ICON_L22_UTC_Time                     (Epoch) object ...
    ICON_L22_Zonal_Wind                   (Epoch, ICON_L22_Altitude) float64 ...
    ICON_L22_Meridional_Wind              (Epoch, ICON_L22_Altitude) float64 ...
    ICON_L22_Zonal_Wind_Error             (Epoch, ICON_L22_Altitude) float64 ...
    ICON_L22_Meridional_Wind_Error        (Epoch, ICON_L22_Altitude) float64 ...
    ICON_L22_Wind_Quality                 (Epoch, ICON_L22_Altitude) float64 ...
    ICON_L22_Fringe_Amplitude             (Epoch, ICON_L22_Altitude) float64 ...
    ICON_L22_Fringe_Amplitude_Error       (Epoch, ICON_L22_Altitude) float64 ...
    ICON_L22_Relative_VER                 (Epoch, ICON_L22_Altitude) float64 ...
    ICON_L22_Relative_VER_Error           (Epoch, ICON_L22_Altitude) float64 ...
    ICON_L22_VER_Quality                  (Epoch, ICON_L22_Altitude) float64 ...
    ICON_L22_Magnetic_Field_Aligned_Wind  (Epoch, ICON_L22_Altitude) float64 ...
    ICON_L22_Magnetic_Meridional_Wind     (Epoch, ICON_L22_Altitude) float64 ...
    ICON_L22_Magnetic_Zonal_Wind          (Epoch, ICON_L22_Altitude) float64 ...
    ICON_L22_Longitude                    (Epoch, ICON_L22_Altitude) float64 ...
    ICON_L22_Latitude                     (Epoch, ICON_L22_Altitude) float64 ...
    ICON_L22_Magnetic_Latitude            (Epoch, ICON_L22_Altitude) float64 ...
    ICON_L22_Magnetic_Longitude           (Epoch, ICON_L22_Altitude) float64 ...
    ICON_L22_Solar_Zenith_Angle           (Epoch, ICON_L22_Altitude) float64 ...
    ICON_L22_Local_Solar_Time             (Epoch, ICON_L22_Altitude) float64 ...
    ICON_L22_Time_Delta                   (Epoch, ICON_L22_Altitude) float64 ...
    ICON_L22_Fringe_Amplitude_A           (Epoch, ICON_L22_Altitude) float64 ...
    ICON_L22_Fringe_Amplitude_B           (Epoch, ICON_L22_Altitude) float64 ...
    ICON_L22_Relative_VER_A               (Epoch, ICON_L22_Altitude) float64 ...
    ICON_L22_Relative_VER_B               (Epoch, ICON_L22_Altitude) float64 ...
    ICON_L22_VER_Relative_Difference      (Epoch, ICON_L22_Altitude) float64 ...
    ICON_L22_Orbit_Number                 (Epoch, ICON_L22_Altitude) float64 ...
    ICON_L22_Orbit_Node                   (Epoch, ICON_L22_Altitude) float64 ...
    ICON_L22_Quality_Flags                (Epoch, ICON_L22_Altitude, N_Flags) float64 ...
```

My pysatNASA branch is `meta_update` and I copied your loading instructions exactly.  Is this expected?

Using the new data keys for the examples, I tried:

```
iloc = np.where(inst.data['ICON_L22_Wind_Quality'] != 1)
inst[iloc, 'ICON_L22_Wind_Quality'] = np.nan
```

Which throws a key error.  When I try the equivalent of your test:

```
inst[iloc[0], iloc[1], 'ICON_L22_Wind_Quality'] = np.nan
```

My computer hangs after running out of memory trying to do this.

aburrell commented:
Ok, since my issues were really with pysatNASA, I used JRO ISR.  I also added a fix for the bug I found.  To test:

```
import datetime as dt
import numpy as np
import pysat
import pysatMadrigal as py_mad

jro = pysat.Instrument(inst_module=py_mad.instruments.jro_isr, tag='drifts')
jro.download(dt.datetime(2010, 1, 18), dt.datetime(2010, 1, 19), user=username, password=email)
jro.load(2010, 18)
iloc = np.where(jro['vipn2'] < 0.0). # Get negative velocities
tst1 = jro[iloc[0], iloc[1], 'vipn2']
tst2 = jro[iloc, 'vipn2']
```

`tst1` and `tst2` will be the same, unless the values are padded (np.nan).  This confirms the @jklenzing fix works and the new fix I added works too.

jklenzing commented:
Thanks @aburrell!  I've updated the testing procedure to check the `getitem` for xarray as well.  Note that some of the original tests were redundant, so I've cleaned that up and expanded it to be consistent with the `setitem` tests.

rstoneback commented:
Thanks @jklenzing and @aburrell. I've run @aburrell's example and it works in this branch (tst1 == tst2) but errors out in the regular develop-3 branch.

I have to prep for an impeding class lecture but I plan on coming back later today to complete the review.

jklenzing commented:
Ready for review.  There's a small decrease in coverage for the `getitem` tests that I don't fully understand.  Any input would be appreciated.

rstoneback commented:
> Ready for review. There's a small decrease in coverage for the `getitem` tests that I don't fully understand. Any input would be appreciated.

I spent some time on the gap in coverage. I believe I identified the issue (the get item input needs to be a multidimensional array) but the tests I created below aren't quite working for me. I'm also failing on some established tests so perhaps my xarray is too old? What version are you using?

These are intended for `TestBasics2DXarray`
```
    @pytest.mark.parametrize("index",
                             [([0, 1, 2], slice(1, None)),
                              ([0, 1, 2], [0, 1, 2]),
                              [[0, 1, 2], [0, 1, 2]],
                              [slice(1, 3), slice(1, 3)]])
    def test_getting_partial_data_by_2d_indices_and_name(self, index):
        """Check data access using multidiminesional index type"""
        self.testInst.load(self.ref_time.year, self.ref_doy)
        # Get data via pysat
        data = self.testInst[index, 'profiles']
        # Get data via chained access
        alt_data = self.testInst['profiles'][index]
        assert np.all(data == alt_data)

    @pytest.mark.parametrize("index",
                             [([0, 1, 2], slice(1, None), [0, 1, 2, 4]),
                              ([0, 1, 2], [0, 1, 2], [0, 1, 2, 4], [5, 6, 7, 8]),
                              [[0, 1, 2], [0, 1, 2], [0, 1, 2, 4]],
                              [slice(1, 3), slice(1, 3), [0, 1, 2, 4], [5, 6]]])
    def test_getting_partial_data_error_by_2d_indices_and_name(self, index):
        """Check data index error check, using multidiminesional index type"""
        self.testInst.load(self.ref_time.year, self.ref_doy)
        # Try to get data via pysat
        with pytest.raises(ValueError) as err:
            data = self.testInst[index, 'profiles']


```

I have to switch to some cosmic and icon stuff for a bit...

rstoneback commented on pysat/tests/test_instrument.py:
```diff
@@ -2405,6 +2361,54 @@ def teardown(self):
+        assert np.all(np.all(self.testInst[changed, changed, 'doubleProfile']
+                             == 0))
+
+
```
```suggestion
    @pytest.mark.parametrize("index",
                             [([0, 1, 2], slice(1, None)),
                              ([0, 1, 2], [0, 1, 2])])
    def test_getting_partial_data_by_2d_indices_and_name(self, index):
        """Check data access using multidiminesional index type"""
        self.testInst.load(self.ref_time.year, self.ref_doy)
        # Get data via pysat
        data = self.testInst[index, 'profiles']
        # Get data via chained access
        alt_data = self.testInst['profiles'][index]
        assert np.all(data == alt_data)

    @pytest.mark.parametrize("index",
                             [([0, 1, 2], slice(1, None), [0, 1, 2, 4]),
                              ([0, 1, 2], [0, 1, 2], [0, 1, 2, 4], [5, 6])])
    def test_getting_partial_data_error_by_2d_indices_and_name(self, index):
        """Check data index error check, using multidiminesional index type"""
        self.testInst.load(self.ref_time.year, self.ref_doy)
        # Try to get data via pysat
        with pytest.raises(ValueError) as err:
            data = self.testInst[index, 'profiles']

```

rstoneback commented:
The problem was my list inputs. I've updated the test code.

aburrell commented:
I also fixed it in a different way.  Should get all the new lines.

aburrell commented:
Also, my internet is rubbish today, so... may or may not be contactable.

aburrell commented:
And since I've done a fair bit of code here, I think @rstoneback should give the approval.

jklenzing commented:
All new tets are running locally.  @rstoneback, I'm using xarray 0.16.0.

rstoneback reviewed (approved):
Thanks @jklenzing and @aburrell.
